instaclustr/resource_cluster.go

```diff
@@ -1542,10 +1584,22 @@ func resourceClusterDelete(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*Config).Client
 	id := d.Get("cluster_id").(string)
 	log.Printf("[INFO] Deleting cluster %s.", id)
+
+	return resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
+		return clusterDeleteRetry(client, d, id)
+	})
+}
+
+func clusterDeleteRetry(client APIClientInterface, d resourceDataInterface, id string) *resource.RetryError {
+	fmt.Printf("waiting for the dependencies of cluster %s to be cleaned\n", id)
+	// try to delete until successful or the error is not about dependency
 	err := client.DeleteCluster(id)
-	if err != nil {
-		return fmt.Errorf("[Error] Error deleting cluster: %s", err)
+	if err != nil && strings.HasPrefix(err.Error(), "Status code: 409") {
+		return resource.RetryableError(fmt.Errorf("[Error] Dependencies still exist for cluster %s", id))
```

Comment:
Just a question here, how does this retry work?

Reply:
Apparently based on the API code, it will return code 409 if there are still dependencies. Now the API client code will wrap the error with string `Status code: X`. This means, we can check if there are still dependencies if the string match.